netlogo-gui/src/main/app/AppTabManager.scala

```diff
@@ -0,0 +1,340 @@
+// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo
+
+package org.nlogo.app
+
+import java.awt.Component
+import java.awt.event.{ ActionEvent, KeyEvent }
+import javax.swing.{ Action, AbstractAction, ActionMap, InputMap, JComponent, JTabbedPane, KeyStroke }
+
+import org.nlogo.app.codetab.{ CodeTab }
+import org.nlogo.swing.{ TabsMenu, UserAction }, UserAction.MenuAction
+
+// The class AppTabManager handles relationships between tabs (the
+// InterfaceTab, the InfoTab, the MainCodeTab and the included files tabs )
+// and their containing classes - Tabs and CodeTabsPanel.
+// When the MainCodeTab is in a separate window the CodeTabsPanel exists
+// and contains all the CodeTabs, while the other tabs belong to Tabs
+// (Tabs and CodeTabsPanel both being JTabbedPanes).
+// Otherwise all the tabs belong to Tabs.
+// AppTabManager uses the variable codeTabsPanelOption of type Option[CodeTabsPanel]
+// to deal with the fact that CodeTabsPanel is instatiated only some of the time.
+
+class AppTabManager( val appTabsPanel:          Tabs,
+                     var codeTabsPanelOption: Option[CodeTabsPanel]) {
+
+  // The appTabsPanel and the main code tab are unique unchanging entities
+  // of class Tabs and MainCodeTab respectively
+  def getAppTabsPanel = { appTabsPanel }
+  def getMainCodeTab = { appTabsPanel.getMainCodeTab }
+
+  // The separate window and JTabbedPane containing the main code tab and
+  // other code tabs can come in an out of existence, and are hence
+  // represented by a scala Option, which has value 'None' when code tabs
+  // are in the Application window and JTabbedPane
+  def setCodeTabsPanelOption(_codeTabsPanelOption: Option[CodeTabsPanel]): Unit = {
+    codeTabsPanelOption = _codeTabsPanelOption
+  }
+
+  // might want access to these owner methods to be only in the app package
+  // Need to carefully decide which methods are private
+  def getCodeTabsOwner = {
+    codeTabsPanelOption match {
+      case None           => appTabsPanel
+      case Some(theValue) => theValue
+    }
+  }
+
+  def getAppTabsOwner = { appTabsPanel }
```

Comment:
I have some get functions such as getAppTabsOwner that are used by other classes to access members of a class.
In this case AbstractTabsPanel uses appTabsPanel. 
As you noted elsewhere, it might be possible to simplify some of this, but the fact that many objects in NetLogo have their state changed in multiple steps after they are first constructed makes things quite ticklish.

Reply:
I'm still not clear on how it's useful to expose `appTabsPanel` as a method in addition to it being a `val` on the class (which will never change), but now `getAppTabsPanel` has also been added which again seems to do the exact same thing?  At a minimum there should only be one of these methods, and I think just letting people access `appTabsPanel` directly is fine.